sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -590,54 +598,81 @@ ProgramManager::ProgramManager() {
                           PI_INVALID_VALUE);
     File.seekg(0, std::ios::end);
     size_t Size = File.tellg();
-    std::unique_ptr<unsigned char[]> Data(new unsigned char[Size]);
+    std::unique_ptr<char[]> Data(new char[Size]);
     File.seekg(0);
-    File.read(reinterpret_cast<char *>(Data.get()), Size);
+    File.read(Data.get(), Size);
     File.close();
     if (!File.good())
       throw runtime_error(std::string("read from ") + SpvFile +
                               std::string(" failed"),
                           PI_INVALID_VALUE);
+    auto ImgPtr = make_unique_ptr<DynRTDeviceBinaryImage>(
+        std::move(Data), Size, OSUtil::DummyModuleHandle);
 
-    std::unique_ptr<DeviceImage, ImageDeleter> ImgPtr(new DeviceImage(),
-                                                      ImageDeleter());
-    ImgPtr->Version = PI_DEVICE_BINARY_VERSION;
-    ImgPtr->Kind = PI_DEVICE_BINARY_OFFLOAD_KIND_SYCL;
-    ImgPtr->DeviceTargetSpec = PI_DEVICE_BINARY_TARGET_UNKNOWN;
-    ImgPtr->CompileOptions = "";
-    ImgPtr->LinkOptions = "";
-    ImgPtr->ManifestStart = nullptr;
-    ImgPtr->ManifestEnd = nullptr;
-    ImgPtr->BinaryStart = Data.release();
-    ImgPtr->BinaryEnd = ImgPtr->BinaryStart + Size;
-    ImgPtr->EntriesBegin = nullptr;
-    ImgPtr->EntriesEnd = nullptr;
-    // TODO the environment variable name implies that the only binary format
-    // it accepts is SPIRV but that is not the case, should be aligned?
-    ImgPtr->Format = getFormat(*ImgPtr);
-
+    if (DbgProgMgr > 0) {
+      std::cerr << "loaded device image binary from " << SpvFile << "\n";
+      std::cerr << "format: " << getFormatStr(ImgPtr->getFormat()) << "\n";
+    }
     // No need for a mutex here since all access to these private fields is
     // blocked until the construction of the ProgramManager singleton is
     // finished.
     m_DeviceImages[SpvFileKSId].reset(
-        new std::vector<DeviceImage *>({ImgPtr.get()}));
+        new std::vector<RTDeviceBinaryImageUPtr>());
+    m_DeviceImages[SpvFileKSId]->push_back(std::move(ImgPtr));
+  }
+}
 
-    m_OrphanDeviceImages.emplace_back(std::move(ImgPtr));
+void ProgramManager::populateSpecConstRegistry() {
+  if (DbgProgMgr > 1) {
+    std::cerr << ">>> ProgramManager::populateSpecConstRegistry\n";
+  }
+  for (const auto &Entry : m_DeviceImages) {
+    const std::vector<RTDeviceBinaryImageUPtr> &Imgs = *Entry.second.get();
 
-    if (DbgProgMgr > 0)
-      std::cerr << "loaded device image from " << SpvFile << "\n";
+    if (Imgs.size() == 0)
```

Comment:
Minor. How can it happen that Imgs.size() is zero here?

Reply:
Looks like it can't happen - will change to assert.